services/wallet/blockchainstate/blockchainstate.go

```diff
@@ -1,63 +1,160 @@
 package blockchainstate
 
+//go:generate go tool mockgen -package=mock_blockchainstate -source=blockchainstate.go -destination=mock/blockchainstate.go
+
 import (
 	"context"
+	"errors"
+	"math"
+	"strconv"
 	"sync"
 	"time"
 
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+
+	"github.com/status-im/go-wallet-sdk/pkg/contracts/multicall3"
+
+	"github.com/status-im/status-go/rpc/chain"
 	"github.com/status-im/status-go/services/wallet/common"
 )
 
+type EthClientGetter interface {
+	EthClient(chainID uint64) (chain.ClientInterface, error)
+}
+
 type LatestBlockData struct {
 	blockNumber   uint64
 	timestamp     time.Time
 	blockDuration time.Duration
 }
 
 type BlockChainState struct {
-	blkMu              sync.RWMutex
+	ethClientGetter    EthClientGetter
+	mu                 sync.RWMutex
 	latestBlockNumbers map[uint64]LatestBlockData
 	sinceFn            func(time.Time) time.Duration
 }
 
-func NewBlockChainState() *BlockChainState {
+func NewBlockChainState(ethClientGetter EthClientGetter) *BlockChainState {
 	return &BlockChainState{
-		blkMu:              sync.RWMutex{},
+		ethClientGetter:    ethClientGetter,
+		mu:                 sync.RWMutex{},
 		latestBlockNumbers: make(map[uint64]LatestBlockData),
 		sinceFn:            time.Since,
 	}
 }
 
+// Get the estimated latest block number for a chain
 func (s *BlockChainState) GetEstimatedLatestBlockNumber(ctx context.Context, chainID uint64) (uint64, error) {
-	blockNumber, _ := s.estimateLatestBlockNumber(chainID)
-	return blockNumber, nil
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	return s.estimateLatestBlockNumber(ctx, chainID)
 }
 
-func (s *BlockChainState) SetLastBlockNumber(chainID uint64, blockNumber uint64) {
-	blockDuration, found := common.AverageBlockDurationForChain[common.ChainID(chainID)]
-	if !found {
-		blockDuration = common.AverageBlockDurationForChain[common.ChainID(common.UnknownChainID)]
+// Get the estimated block time for a given block number and a chain
+func (s *BlockChainState) GetEstimatedBlockTime(ctx context.Context, chainID uint64, blockNumber uint64) (time.Time, error) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	return s.estimateBlockTime(ctx, chainID, blockNumber)
+}
+
+// Set the latest known block number for a chain
+func (s *BlockChainState) SetLatestBlockNumber(chainID uint64, blockNumber uint64) {
+	s.mu.RLock()
```

Comment:
I suspect we'd better use a `Lock()` here, because `setLatestBlockNumber` modifies the state

Reply:
oops, nice catch!